### PR TITLE
feat:add cache for the g(_, _) function.

### PR DESCRIPTION
**Pull request by M1serry**

### CODE REVIEW

**Comment by hsluoyz** [comment]
@leeqvip

**Comment by leeqvip** [comment]
@M1serry Some checks were not successful, fix please.

**Review by leeqvip — CHANGES_REQUESTED** [review]
Unit test failed, please fix

**Review by leeqvip — CHANGES_REQUESTED** [review]

File: src/Util/BuiltinOperations.php [hunk 1]
```diff
@@ -430,18 +430,33 @@ public static function globMatchFunc(...$args): bool
...
+            $key = '';
+            foreach ($args as $arg) {
+                $key .= (string)$arg;
+            }
+
+            foreach ($memorized as $k => $v) {
+                if ($k == $key) {
+                    return $v;
+                }
+            }
```
[the remark this answers is not in the document]
Comment (leeqvip) [reply src/Util/BuiltinOperations.php:444]: @M1serry See suggested change .

File: src/Util/BuiltinOperations.php [hunk 2]
```diff
@@ -431,17 +431,30 @@ public static function globMatchFunc(...$args): bool
     public static function generateGFunction(RoleManager $rm = null): Closure
     {
         return function (...$args) use ($rm) {
+            $memorized = [];
```
Comment (leeqvip) [inline src/Util/BuiltinOperations.php:434]: ` $memorized` should be defined outside.

**Review by leeqvip — CHANGES_REQUESTED** [review]

File: src/Util/BuiltinOperations.php [hunk 3]
```diff
@@ -430,18 +430,31 @@ public static function globMatchFunc(...$args): bool
...
     public static function generateGFunction(RoleManager $rm = null): Closure
     {
-        return function (...$args) use ($rm) {
+        $memorized = [];
+        return function (...$args) use ($rm, $memorized) {
+            $key = '';
+            foreach ($args as $arg) {
+                $key .= (string)$arg;
+            }
+
```
Comment (leeqvip) [inline src/Util/BuiltinOperations.php:439]:
```suggestion
            /** @phpstan-ignore-next-line */
```

**Pull request merged**